ibis/expr/rules.py

```diff
@@ -794,3 +794,81 @@ def _validate(self, args, i):
 
 def comparable(left, right):
     return ir.castable(left, right) or ir.castable(right, left)
+
+
+class Table(ValueTyped):
+    """A table argument.
+
+    Parameters
+    ----------
+    name : str
+        The name of the table argument.
+    optional : bool
+        Whether this table argument is optional or not.
+    satisfying : iterable
+        An iterable of lambda expressions and/or column rules. These will be
+        used to validate arguments. Lambda expressions must take a table
+        as their single argument, validate, and return ``True`` (validated
+        successfully) or ``False`` (not validated). For column rules, if at
+        least one column in the table matches the rule, then the table
+        is valid.
+
+    validator : ???
+        ???
+    doc : ???
+        ???
+
+    Examples
+    --------
+    The following op will accept an argument named ``'table'``, with at least
+    four columns, of which at least three must be of type ``double``. The names
+    and types of each column must also match those in the rule (e.g. column
+    with name ``'value4'`` is an optional column but must be of type
+    ``double``).
+
+    >>> class MyOp(ops.ValueOp):
+    ...    input_type = [
+    ...        rules.table(
+    ...            name='table',
+    ...            satisfying=[
+    ...                lambda t: len(t.columns) >= 4,
+    ...                lambda t: t.schema().types.count(dt.Double()) >= 3,
+    ...                rules.column(name='group', value_type=rules.number),
+    ...                rules.column(name='value1', value_type=rules.number),
+    ...                rules.column(name='value2', value_type=rules.number),
+    ...                rules.column(name='value3', value_type=rules.number),
+    ...                rules.column(name='value4', value_type=rules.number,
+    ...                             optional=True)])]
+    ...    output_type = rules.type_of_arg(0)
+    """
+    def __init__(self, name=None, optional=False, satisfying=None,
+            doc=None, validator=None, **arg_kwds):
+        self.name = name
+        self.optional = optional
+        self.satisfying = util.promote_list(satisfying)
+
+        self.doc = doc
+        self.validator = validator
+
+    def _validate(self, args, i):
```

Comment:
@cpcloud I just realized that with the current implementation a table can pass validation simply by having enough columns to pass the rules. We can of course control the amount of columns with lambda rules, but this might be counterintuitive. Maybe we should have an argument like `allow_extra`, which changes the behavior for allowing columns that do not match a rule.

Reply:
I implemented this.